### DISCUSSION TITLE
Let directories optionally define their own index page (within)

### DISCUSSION THREAD

**Discussion by vcavallo** [srid/emanote#107] [Ideas]
In the current implementation, if you have a directory that you want to have it's own "landing page" upon clicking the directory, you need to define such a file at the root. Illustrating...

## Case 1: A self-contained collection:

```
/some-file.md
/some-other-file.md
/literature-notes/
/literature-notes/some-book.md
/literature-notes/some-podcast.md
```

In the generated emanote site, if you click on the `literature-notes` folder in the sidebar (let's call this page it's "index") you'll see some content along with the notice:

> Note: To override the auto-generated content here, create a file named: literature-notes.md

and it will expect the file to be in the root, like so:

```
/some-file.md
/some-other-file.md
/literature-notes.md                               <----- HERE
/literature-notes/
/literature-notes/some-book.md
/literature-notes/some-podcast.md
```

I propose that it be possible to define this "index file" _within_ the directory itself, so as to avoid polluting the environment outside of itself.  like so:

```
/some-file.md
/some-other-file.md
/literature-notes/
/literature-notes/index.md                     <----- HERE
/literature-notes/some-book.md
/literature-notes/some-podcast.md
```

Or even:

```
/some-file.md
/some-other-file.md
/literature-notes/
/literature-notes/literature-notes.md     <----- HERE
/literature-notes/some-book.md
/literature-notes/some-podcast.md
```

Although I can see a good argument for wanting to have it both ways or even merging the two. see below:

## Case 2: A category that warrants a separate treatment outside of the directory

Imagine you have a personal website that describes some aspects of your life. You might want a "hobbies" file that talks about how you feel about hobbies in general, along with a "hobbies/" directory that lists yours.

```
/about-me.md
/hobbies.md
/hobbies/
/hobbies/index.md
/hobbies/gardening.md
/hobbies/drumming.md
```

In `/hobbies.md` you might talk about research surrounding hobbies. You may also refer to `[my hobbies](/hobbies/)` so a visitor can learn more about yours in particular. `/hobbies/index.md` may give some preamble and then list the members of the directory - entirely different purpose than the root-level `hobbies.md`. 

## Case 3: Merging

Perhaps if you have _both_ `/hobbies.md` _and_ one of `/hobbies/index.md` or `/hobbies/hobbies.md`, then the contents of the root-level `/hobbies.md` should be merged into the generated output of the leaf-level `/hobbies/hobbies/md` somewhere. Perhaps this should be optional.

**Comment by kouloumos** [comment]
This is also useful when trying to take advantage of the different routes functionality in order to have different knowledge-bases (possibly managed by different people) as submodules of a larger knowledge-base. In this case, not having all the information in the individual submodules is not very convenient.

It seems that there is acceptance for this idea; if someone can point me to a direction on where to start, I could possibly try to implement it.

**Reply by srid** [reply]
I'm not sure what the right solution here should be. Obsidian (and many Markdown wikilink extensions) doesn't automatically resolve `[[Test]]` to `./Test/index.md` if it exists. I suppose this will work: have Emanote resolve `[[Test]]` to `./Test/index.md` but do so _only if_ `./Test.md` doesn't already exist. Users have a choice to use one over the other.

> if someone can point me to a direction on where to start, I could possibly try to implement it.

Perhaps start from here: https://github.com/srid/emanote/blob/f240333f9fd60d088b9a5b32632d8bbc996c080c/src/Emanote/Model/Link/Resolve.hs#L13-L23

This is potentially also relevant: https://github.com/srid/emanote/blob/f240333f9fd60d088b9a5b32632d8bbc996c080c/src/Emanote/Model/Type.hs#L249-L254

But you may first want to tweak the 'allowed wiki-links' index here:

https://github.com/srid/emanote/blob/f240333f9fd60d088b9a5b32632d8bbc996c080c/src/Emanote/Model/Note.hs#L54-L55

https://github.com/srid/emanote/blob/f240333f9fd60d088b9a5b32632d8bbc996c080c/src/Emanote/Model/Note.hs#L81-L90

---

General Haskell hacking tips:

- Use VSCode, and `bin/run` is your friend for fast compile feedback cycle.
- Use [`Debug.Trace`](https://hackage.haskell.org/package/base-4.18.0.0/docs/Debug-Trace.html) to trace values throughout the code base

**Comment by srid** [comment]
https://github.com/srid/emanote/issues/494

**Reply by srid** [reply]
Working demo: https://github.com/srid/emanote/pull/512
